File: 0133_cloneGraph/main.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if(!node)
            return nullptr;
        unordered_map<Node*, Node*> visited;

        return newNode(node, visited);
    }

    Node* newNode(Node* node, unordered_map<Node*, Node*>& visited)
    {
        if(visited.find(node) != visited.end())
            return visited[node];
        
        Node* cloneNode = new Node(node->val);
        visited[node]=cloneNode;

        for(int i = 0; i < node->neighbors.size(); ++i)
        {
            cloneNode->neighbors.push_back(newNode(node->neighbors[i], visited));
        }
        return cloneNode;
    }
};

File: 0133_cloneGraph/main.py
from typing import Dict

class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        def newNode(_map: Dict['Node', 'Node'], _node: 'Node' ):
            if node is None:
                return None

            cloneNode = Node(_node.val, [])
            map[_node] = cloneNode

            for neighbor in _node.neighbors:
                if neighbor in map:
                    cloneNode.neighbors.append(map[neighbor])
                else:
                    cloneNode.neighbors.append(newNode(_map, neighbor))
            return cloneNode

        map: Dict['Node', 'Node'] = dict()
        return newNode(map, node)



if __name__ == '__main__':
    sol = Solution()
    sol.cloneGraph(None)

